### PR TITLE
Add class as a target for RequiresUnreferencedCodeAttribute

### PR DESCRIPTION
**Pull request by tlakollo**
## Background and Motivation

As explained in https://github.com/mono/linker/issues/1742 some types are considered not trim compatible leading to mark all its member methods with RequiresUnreferencedCode. This could lead to hundreds of members annotated with RequiresUnreferencedCode. The proposal is to only annotate the type with RequiresUnreferencedCode, then the linker will treat this as if we annotated constructors and static methods with RequiresUnreferencedCode.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/RequiresUnreferencedCodeAttribute.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/RequiresUnreferencedCodeAttribute.cs:14]: Why just class, rather than also, say, struct?

We should review the API change as part of API review, though mainly just to ensure we've properly discussed exactly what targets we want to enable, not because anyone would have a real problem with it.
cc: @terrajobst

Comment (MichalStrehovsky) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/RequiresUnreferencedCodeAttribute.cs:14]: Structs are problematic because they can be default-instantiated. We would like this to be considered unsafe (and warn) only if a constructor or a static method on the type is called to limit the amount of noise (and additional suppressions needed) when a type is annotated as such.

If we allow this on structs, we would have to warn on instance methods too, plus it would introduce problems when unsafe instance methods get called indirectly through interfaces (the tool would not be able to see that at all).

It's likely not worth the trouble.

Comment (tlakollo) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/RequiresUnreferencedCodeAttribute.cs:14]: Sorry for skipping the process, I created https://github.com/dotnet/runtime/issues/50122 to discuss the API change

File: src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt
Comment (eerhardt) [inline src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt:28]: Why is this `18` instead of `17`? Didn't we just add 1 new warning?

File: src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt
Comment (eerhardt) [inline src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt:25]: Does this change have to happen? Or was this generated?

File: src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt
Comment (tlakollo) [reply src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt:28]: I just count the number of warnings after the merge with main

File: src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt
Comment (tlakollo) [reply src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt:25]: It was a conflicting change in main

Comment (eerhardt) [reply src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt:25]: I don't think that answers the question.

Comment (tlakollo) [reply src/libraries/shims/ApiCompatBaseline.PreviousNetCoreApp.txt:25]: Initially, it was generated, I had to make some changes because the count is wrong since a while back, and the conflicting attribute was under `Compat issues with assembly System.Security.Cryptography` instead of `Compat issues with assembly System.Runtime`

**Review by eerhardt — APPROVED** [review]

**Pull request merged**
